src/scores/functions.py

```
@@ -33,6 +34,86 @@ def apply_weights(values, *, weights: Optional[XarrayLike] = None):
+
+    Args:
+        values: Input data to be reduced. Typically an `xr.DataArray` or `xr.Dataset`.
+        reduce_dims: Dimensions over which to apply the mean. Can be a string, list of
+            strings, or None. If None, no reduction is performed. Defaults to None.
+        weights: Weights to apply for weighted averaging.
+            Must be broadcastable to `values` and contain no negative values. If None,
+            an unweighted mean is calculated. Defaults to None.
+        method: Aggregation method to use. Either "mean" or "sum". Defaults to "mean".
+
+    Returns:
+        An xarray object (same type as the input) with (un)weighted mean of ``values``
+
+    Raises:
+        ValueError: If `weights` contains any negative values.
+
+    Warnings:
+        UserWarning: If weights are provided but no reduction is performed (`reduce_dims` is None),
+        a warning is issued since weights are ignored.
+
+    Examples:
+        >>> import xarray as xr
+        >>> import numpy as np
+        >>> da = xr.DataArray(np.arange(6).reshape(2, 3), dims=['x', 'y'])
+        >>> weights = xr.DataArray([1, 2], dims=['x'])
+        >>> apply_weighted_mean(da, reduce_dims=['x'], weights=weights)
+        <xarray.DataArray (y: 3)>
+        array([2., 3., 4.])
+        Dimensions without coordinates: y
+
+    """
+    if method not in ["mean", "sum"]:
+        raise ValueError(f"Method must be either 'mean' or 'sum', got '{method}'")
+
+    if weights is not None:
+        if (weights < 0).any():
+            raise ValueError("Weights must not contain negative values.")
+    if reduce_dims is None and weights is not None:
+        warnings.warn(
+            """
+            Weights were provided but all the score across all dimensions is being preserved. 
+            Weights will be ignored.
+            """,
+            UserWarning,
+        )
+    if reduce_dims is not None:
+        if weights is not None:
+            weights = weights.fillna(0)
```

I think we raise an error or a warning saying its being forced to 0 if its nan. A weight has to be numeric. As you've done here, 0 is fair assumption as it implies exclusion.

However, a user that inputted weights with both 0 and nans, causes ambiguous intent from the code pespective.
- `0`: "I do not want to give this value a weight" = "I want to give this value 0 weight" => "I want to exclude this value from this weight calculation". This is unambiguous.
- `nan`: "The weighting for this value is undefined" or "I do not know what to weight this value with" => "Impute the weight from neighbours?", "Exclude?", "Take the average weight?"

If we have a warning saying its forced to 0 - then its clear to the user, or else, if we want to enforce that `nan => 0` then we raise an error, since in reality `nan != 0`; and request that the user fix it.

---

In general, we calculate results wherever the values are not NaN, so I'm also okay with this approach and think it's convenient. I can see conceivable merit in allowing multiple strategies to dealing with NaNs but in practice in `scores` I think we are always just dropping them. Having a convenience arg parameter of weight_when_nan=0 could be appropriate, and easier to deal with than silencing warnings. That leaves the door open for the user to provide another numeric value, or perhaps one day a strategy (like average or infull). I am also okay with just always treating it as 0 and documenting it clearly.

---

I’d prefer an error over warnings as well. But I would not prefer to stack default args or options if possible. They just get propagated to higher level calls and lead to argument bloat. For example xarray open dataset has so many options that its base functionality gets obscured. Each additional argument is a categorical product leading to exponentially growing number of branch possibilities in terms of code. Which results in an exponential growing number of test coverage. 

It’s preferable to curtail that where there is a possibility of curtailing it. If we do not want to produce warnings when forcing we should throw an exception.